### PR TITLE
Artur Zaprzala's rdp2tcp channel driver

### PR DESCRIPTION
**Pull request by widgetii**
This is Artur Zaprzala's work adapting rdp2tcp protocol from http://rdp2tcp.sourceforge.net/ project
I use it already for several months on both Linux and MacOs machines and it works great.

Discussed in https://github.com/FreeRDP/FreeRDP/issues/935

I feel that many people definitely need this also

### CODE REVIEW

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by widgetii** [comment]
Something with Windows build went wrong? Cannot recognize real error message

**Comment by akallabeth** [comment]
https://ci.freerdp.com/job/windows_msbuild_vs2010/arch=win64,label=vs2010/2022/console

Some types not included or broken casts?
Btw, could you run the formatter script, you have some space/tab mixups and inconsistent bracing styles.

**Comment by widgetii** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@widgetii sorry, only listens to a few :/ @freerdp-bot test ;)

File: channels/rdp2tcp/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# FreeRDP: A Remote Desktop Protocol Implementation
+# FreeRDP cmake build script
+#
+# Copyright 2012 Marc-Andre Moreau <marcandre.moreau@gmail.com>
```
Comment (akallabeth) [inline channels/rdp2tcp/CMakeLists.txt]: sure about that copyright?

File: winpr/libwinpr/file/generic.c [hunk 2]
```diff
@@ -178,6 +178,11 @@
 
 #define EPOCH_DIFF 11644473600LL
 #define STAT_TIME_TO_FILETIME(_t) (((UINT64)(_t) + EPOCH_DIFF) * 10000000LL)
+#ifdef _STAT_VER_LINUX
+#define STAT_TIME_TO_FILETIME2(_t, _tl) (((UINT64)(_tl.tv_sec) + EPOCH_DIFF) * 10000000LL + _tl.tv_nsec/100)
+#else
+#define STAT_TIME_TO_FILETIME2(_t, _tl) STAT_TIME_TO_FILETIME(_t)
```
Comment (akallabeth) [inline winpr/libwinpr/file/generic.c]: this change is not related to the pr, is it?

**Comment by akallabeth** [comment]
@freerdp-bot sometimes does not even test if we tell it to ;)
could you elaborate what the changes in `file.c` are for? they seem to be unrelated to the tcp patch.

File: channels/rdp2tcp/CMakeLists.txt (hunk 1 above)
Comment (widgetii) [reply channels/rdp2tcp/CMakeLists.txt]: I know that the code has long history and believe that this is one of first authors

File: winpr/libwinpr/file/generic.c (hunk 2 above)
Comment (widgetii) [reply winpr/libwinpr/file/generic.c]: definitely yes, removed from PR

File: winpr/libwinpr/file/file.c [hunk 3]
```diff
@@ -463,30 +475,32 @@ static BOOL FileUnlockFileEx(HANDLE hFile, DWORD dwReserved, DWORD nNumberOfByte
...
 		return FALSE;
 	}
-#endif
 
+#endif
 	return TRUE;
 }
 
-static UINT64 FileTimeToUS(const FILETIME* ft)
+static UINT64 FileTimeTo100ns(const FILETIME* ft)
```
Comment (akallabeth) [inline winpr/libwinpr/file/file.c]: This is the only change in here, correct?
Better remove that as it is a change not related to your tcp patch.

**Comment by akallabeth** [comment]
@freerdp-bot test

File: winpr/libwinpr/file/file.c (hunk 3 above)
Comment (widgetii) [reply winpr/libwinpr/file/file.c]: yes
removed and tested, everything is ok

**Comment by widgetii** [comment]
Fixed, now they are both gone

**Comment by akallabeth** [comment]
@freerdp-bot test please
@widgetii thank you, it looks much better now.

**Comment by widgetii** [comment]
Do I need to do something else?

**Comment by akallabeth** [comment]
@widgetii sorry, no you just need to wait.
need a second review before merging (and currently most devs are busy)

**Comment by widgetii** [comment]
Does this job have a chance for merge?

**Comment by akallabeth** [comment]
@hardening @bmiklautz what do you think?

File: channels/rdp2tcp/client/rdp2tcp_main.c [hunk 4]
```diff
@@ -0,0 +1,314 @@
...
+	saAttr.bInheritHandle = TRUE;
+	saAttr.lpSecurityDescriptor = NULL;
+	siStartInfo.cb = sizeof(STARTUPINFO);
+	siStartInfo.hStdError = GetStdHandle(STD_ERROR_HANDLE);
+	siStartInfo.dwFlags = STARTF_USESTDHANDLES;
+
+	// Create pipes
+	if (!CreatePipe(&plugin->hStdOutputRead, &siStartInfo.hStdOutput, &saAttr, 0))
+	{
+		perror("stdout CreatePipe");
```
Comment (hardening) [inline channels/rdp2tcp/client/rdp2tcp_main.c]: Could you change all the `perror` occurences and replace then with `WLog_ERR` ?

File: channels/rdp2tcp/client/rdp2tcp_main.c [hunk 5]
```diff
@@ -0,0 +1,314 @@
...
+
+static void VCAPITYPE VirtualChannelOpenEventEx(LPVOID lpUserParam, DWORD openHandle, UINT event,
+        LPVOID pData, UINT32 dataLength, UINT32 totalLength, UINT32 dataFlags)
+{
+	Plugin* plugin = (Plugin*)lpUserParam;
+
+	switch (event)
+	{
+		case CHANNEL_EVENT_DATA_RECEIVED:
+			;
```
Comment (hardening) [inline channels/rdp2tcp/client/rdp2tcp_main.c]: extra semicolon

**Comment by akallabeth** [comment]
@freerdp-bot test

**Comment by akallabeth** [comment]
@hardening @bmiklautz anything preventing a merge?

**Comment by widgetii** [comment]
Hi, are there any updates regarding this?

**Comment by akallabeth** [comment]
@widgetii took some time, but finally ;)
Thank you for your work!

**Comment by widgetii** [comment]
Thank you for your help too!

**Comment by elmosgot** [comment]
I miss some documentation how to setup this plugin

**Comment by widgetii** [comment]
FreeRDP contains only adapter for multiplexing tcp connections inside RDP protocol. You'll need to build [rdp2tcp](https://github.com/V-E-O/rdp2tcp) tool for both server (runs on Windows server) and client sides (runs on same machine as FreeRDP). Fell free to read [rdp2tcp docs](https://raw.githubusercontent.com/V-E-O/rdp2tcp/master/README), they are a little bit outdated but comprehensible.

- you'll need to copy rdp2tcp Windows executable to Windows server at first and run it (I hardcoded infinite loop into it and placed to autorun because to don't want classical RDP desktop, only connect hosts behind Windows server (moreover hiding FreeRDP window using Xvfb)
- add new parameter to FreeRDP like this `/rdp2tcp:$HOME/git/rdp2tcp/client/rdp2tcp`, make connection to RDP server and ensure message `virtual channel connected` appeared in console
- run `~/git/rdp2tcp/tools/rdp2tcp.py add forward 127.0.0.1 10001 10.65.102.20 22`, in this case `127.0.0.1 10001` ip address and port on local machine, and `10.65.102.20 22` ip address and port of host behind Windows. Run this after making RDP connection otherwise it will fail.
- connect to local tunnel endpoint like this `ssh -p 10001 127.0.0.1`

Does it make sense?

P.S. I have no enough time to make a PR for FreeRDP doc, it could be valuable if somebody makes this one based on his setup and troubleshooting experience.

**Comment by gerardo-junior** [comment]
I had some problems to execute, so I wrote a step by step:

https://github.com/gerardo-junior/rdp2tcp/blob/master/README.md

**Comment by akallabeth** [comment]
@gerardo-junior feel free to add this to https://github.com/FreeRDP/FreeRDP/wiki
